plugins/outputs/azuremonitor/azuremonitor.go

```diff
@@ -0,0 +1,445 @@
+	urlTemplate            string = "https://%s.monitoring.azure.com%s/metrics"
+	resourceIDTemplate     string = "/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Compute/virtualMachines/%s"
+	vmInstanceMetadataURL  string = "http://169.254.169.254/metadata/instance?api-version=2017-12-01"
+	msiInstanceMetadataURL string = "http://169.254.169.254/metadata/identity/oauth2/token"
+)
+
+var sampleConfig = `
+  ## The resource ID against which metric will be logged.  If not
+  ## specified, the plugin will attempt to retrieve the resource ID
+  ## of the VM via the instance metadata service (optional if running 
+  ## on an Azure VM with MSI)
+  #resource_id = "/subscriptions/<subscription_id>/resourceGroups/<resource_group>/providers/Microsoft.Compute/virtualMachines/<vm_name>"
+  ## Azure region to publish metrics against.  Defaults to eastus.
+  ## Leave blank to automatically query the region via MSI.
+  #region = "useast"
+
+  ## Write HTTP timeout, formatted as a string.  If not provided, will default
+  ## to 5s. 0s means no timeout (not recommended).
+  # timeout = "5s"
+
+  ## Whether or not to use managed service identity.
+  #use_managed_service_identity = true
+
+  ## Fill in the following values if using Active Directory Service
+  ## Principal or User Principal for authentication.
+  ## Subscription ID
+  #azure_subscription = ""
+  ## Tenant ID
+  #azure_tenant = ""
+  ## Client ID
+  #azure_client_id = ""
+  ## Client secrete
+  #azure_client_secret = ""
+`
+
+// Description provides a description of the plugin
+func (a *AzureMonitor) Description() string {
+	return "Configuration for sending aggregate metrics to Azure Monitor"
+}
+
+// SampleConfig provides a sample configuration for the plugin
+func (a *AzureMonitor) SampleConfig() string {
+	return sampleConfig
+}
+
+// Connect initializes the plugin and validates connectivity
+func (a *AzureMonitor) Connect() error {
+	a.client = &http.Client{
+		Transport: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+		},
+		Timeout: a.Timeout.Duration,
+	}
+
+	// If no direct AD values provided, fall back to MSI
+	if a.AzureSubscriptionID == "" && a.AzureTenantID == "" && a.AzureClientID == "" && a.AzureClientSecret == "" {
+		a.useMsi = true
+	} else if a.AzureSubscriptionID == "" || a.AzureTenantID == "" || a.AzureClientID == "" || a.AzureClientSecret == "" {
+		return fmt.Errorf("E! Must provide values for azure_subscription, azure_tenant, azure_client and azure_client_secret, or leave all blank to default to MSI")
+	}
+
+	if !a.useMsi {
+		// If using direct AD authentication create the AD access client
+		oauthConfig, err := adal.NewOAuthConfig(azure.PublicCloud.ActiveDirectoryEndpoint, a.AzureTenantID)
+		if err != nil {
+			return fmt.Errorf("E! Could not initialize AD client: %s", err)
+		}
+		a.oauthConfig = oauthConfig
+	}
+
+	// Pull region and resource identifier
+	err := a.GetInstanceMetadata()
+	if err != nil && a.ResourceID == "" && a.Region == "" {
```

Comment:
I think this should be `||` instead of `&&`, we should always return `if err != nil`, so that we are sure to see all errors and I think we need both of these set.

Also, if Region and ResourceID are set beforehand, maybe we can skip this function completely?